src/back/spv/writer.rs

```diff
@@ -1318,6 +1400,494 @@ impl Writer {
+                    }
+                }
+
+                self.get_index_constant(known_max_index)?
+            }
+            KnownOrComputed::Computed(max_index_id) => max_index_id,
+        };
+
+        // One or the other of the index or length is dynamic, so emit code for
+        // IndexBoundsCheckPolicy::Restrict.
+        let restricted_index_id = self.id_gen.next();
+        block.body.push(Instruction::ext_inst(
+            self.gl450_ext_inst_id,
+            spirv::GLOp::UMin,
+            self.get_uint_type_id()?,
+            restricted_index_id,
+            &[index_id, max_index_id],
+        ));
+        Ok(KnownOrComputed::Computed(restricted_index_id))
+    }
+
+    /// Write an index bounds comparison to `block`, if needed.
+    ///
+    /// If we're able to determine statically that `index` is in bounds for
+    /// `sequence`, return `Known(value)`, where `value` is the actual value of
+    /// the index. (In principle, one could know that the index is in bounds
+    /// without knowing its specific value, but in our simple-minded situation,
+    /// we always know it.)
+    ///
+    /// If instead we must generate code to perform the comparison at run time,
+    /// return `Computed(comparison_id)`, where `comparison_id` is an
+    /// instruction producing a boolean value that is true if `index` is in
+    /// bounds for `sequence`.
+    ///
+    /// The `sequence` expression may be a `Vector`, `Matrix`, or `Array`, a
+    /// `Pointer` to any of those, or a `ValuePointer`. An array may be
+    /// fixed-size, dynamically sized, or use a specializable constant as its
+    /// length.
+    #[allow(clippy::too_many_arguments)]
+    fn write_index_comparison(
+        &mut self,
+        sequence: Handle<crate::Expression>,
+        index: Handle<crate::Expression>,
+        ir_module: &crate::Module,
+        ir_function: &crate::Function,
+        fun_info: &FunctionInfo,
+        function: &mut Function,
+        block: &mut Block,
+    ) -> Result<KnownOrComputed<u32>, Error> {
+        let index_id = self.cached[index];
+
+        // Get the sequence's length. Return early if we've already done the
+        // bounds check.
+        let length_id = match self.write_sequence_length(
+            sequence,
+            ir_module,
+            ir_function,
+            fun_info,
+            function,
+            block,
+        )? {
+            KnownOrComputed::Known(known_length) => {
+                // When the length and index are known at compile time, we do the bounds check
+                // during validation, so there's no need to check anything here.
+                // Mimic the logic in `Validator::validate_expression`.
+                if let crate::Expression::Constant(index_k) = ir_function.expressions[index] {
+                    if let Some(known_index) = ir_module.constants[index_k].to_array_length() {
+                        return Ok(KnownOrComputed::Known(known_index));
```

Comment:
I'm confused about this. What is the expected type returned here? I'd think it's a boolean, i.e. the function returns something that is a boolean index comparison. But this branch returns an index. How so?